### DISCUSSION TITLE
[with-docker] why run yarn install after build?

### DISCUSSION THREAD

**Discussion by millievn** [vercel/next.js#14449] [Help] [answered]
I find [Dockerfile.multistage](https://github.com/vercel/next.js/blob/canary/examples/with-docker/Dockerfile.multistage) and see:
```dockerfile
RUN yarn install
COPY . .
RUN yarn build && yarn --production
```

Why run yarn install after build? Why not run as below?

```dockerfile
RUN yarn --frozen-lockfile --production
COPY . .
RUN yarn build
```
@areai51 @tomevans18 @evenchange4 @JeromeFitz

**Comment by JeromeFitz** [comment, marked as answer]
Hello @qiwang97  😁 

The build process required `devDependencies` and to slim down the copy image that comes after they were removed post build.

Examples could always use a fresh approach, so if you see room for improvement we can continue the conversation here and work towards a PR. Admittedly, I've gone away from Docker myself. 🐳

**Reply by millievn** [reply]
Sorry to say, I misunderstand what `yarn --production` does. It totally covers node_modules folder and removes all devDependencies packages. 👀

**Reply by aizada92** [reply]
Guys I am facing error while even adding RUN yarn install --production I just want to see only prod dependencies. Any suggestions ? I am using multi stage dockerfile

**Reply by millievn** [reply]
> Guys I am facing error while even adding RUN yarn install --production I just want to see only prod dependencies. Any suggestions ? I am using multi stage dockerfile

No enough information to analyse. But if your production build fails with `xxx modules not found` error, maybe the necessary dependencies are placed in devDependencies by mistake.

**Comment by areai51** [comment]
Hi @qiwang97 you have a valid point, I also noticed the docker build was failing because it had a step to copy the next.config file which wasn't present.. here is the PR with some improvements  https://github.com/vercel/next.js/pull/14513.
I've refrained from using `RUN yarn --frozen-lockfile --production` because for people who try to run the docker steps directly without doing a local install may not have the lockfile.

**Comment by millievn** [comment]
@areai51  By the way, is there any way to add some runtime variables when run by docker images？
I did this：
```ts
// client.ts
import axios from "axios";

export default axios.create({
  baseURL: process.env.NEXT_PUBLIC_API,
  timeout: 1000,
});
```

```ts
// .env.development
NEXT_PUBLIC_API = "https://dev.github.com"
```

```Dockerfile
// ... Dockerfile like you said
FROM mhart/alpine-node:base
WORKDIR /app
ARG NEXT_PUBLIC_API=${NEXT_PUBLIC_API}
ENV NEXT_PUBLIC_API=${NEXT_PUBLIC_API}
// ...
CMD ["node_modules/.bin/next","start"]
```

```bash
docker build -t test .
docker run --rm  -e "NEXT_PUBLIC_API=https://prod.github.com" -p 3000:3000 --name test-container test
```

But failure for `process.env.NEXT_PUBLIC_API` is replaced by [next-build](https://github.com/vercel/next.js/blob/canary/packages/next/cli/next-build.ts) in building process.
